### DISCUSSION TITLE
Better Error Definitions

### DISCUSSION THREAD

**Discussion by benedictjohannes** [mikro-orm/mikro-orm#4987] [General]
**Is your feature request related to a problem? Please describe.**
I would like a more complete error handling in mikro-orm.

I found that using _if-else_ on `instanceof err` is one of the most common way to detect what cause an error throughout a _try-catch_ block.  However, with so many errors encapsulated through _static methods_ of only 4 error clasess _(ValidationError, OptimisticLockError, MetadataError, NotFoundError)_, that approach is very limiited. As such, to appropriately identify errors, an error message string matching would be needed.

**Describe the solution you'd like**
I would suggest that
- Each static method of error classes (ie, `NotFoundError.findOneFailed` and `NotFoundError.findExactlyOneFailed` being expanded to its own classes that extend `NotFoundError`). 
- As exporting static method would result in a lot of errors being exported, only the parent error classes (the currently exported error classes) should be exported from `@mikro-orm/core` while the detailed error classes (ie `FindOneNotFoundError`) would be exported  from a separate file. 
- This would allow various error handling approaches 
[such as ones discussed here](https://stackoverflow.com/q/36332665)

**Describe alternatives you've considered**
I try to consider how errors can be encapsulated the closer to the _(IMHO ideal)_ way errors is handled in Go. My proposed solution is the closest I can think of.

**Additional context**
One of the thing I missed most after coding in Go is a more robust error handling. Go code can use `errors.Wrap()` (to rethrow more general errors) and `errors.Is` (to identify type of error in the error chain).

**Comment by B4nan** [comment]
> However, with so many errors encapsulated through static methods of only 4 error clasess (ValidationError, OptimisticLockError, MetadataError, NotFoundError), that approach is very limiited

FYI there are many other errors that can be thrown by making a bad query, e.g. https://github.com/mikro-orm/mikro-orm/blob/master/packages/postgresql/src/PostgreSqlExceptionConverter.ts

Also, you can already customize the `findOneOrFailed` error yourself. I am not sure if people really use both the exact and default approaches, I'd say they pick one behaviour they want, so I don't think there is a real need to distinguish those two.
